src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1371,21 +1378,29 @@ private bool ShouldAsyncBeTreatedAsModifier(bool parsingStatementNotDeclaration)
 
         private static bool IsNonContextualModifier(SyntaxToken nextToken)
         {
-            return GetModifier(nextToken) != DeclarationModifiers.None && !SyntaxFacts.IsContextualKeyword(nextToken.ContextualKind);
+            return GetModifier(nextToken.Kind, contextualKind: SyntaxKind.None) != DeclarationModifiers.None;
         }
 
-        private bool IsPartialType()
+        private static bool CanFollowPartialKeywordOfPartialType(SyntaxToken token, out bool isPartialFollowedByType)
         {
-            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword);
-            switch (this.PeekToken(1).Kind)
+            switch (token.Kind)
             {
                 case SyntaxKind.StructKeyword:
                 case SyntaxKind.ClassKeyword:
                 case SyntaxKind.InterfaceKeyword:
+                    isPartialFollowedByType = true;
                     return true;
+
+                default:
+                    isPartialFollowedByType = false;
+                    return IsNonContextualModifier(token);
```

Comment:
>return IsNonContextualModifier(token); [](start = 20, length = 38)

It feels like we should ensure that there is a struct/class/interface token ahead before returning true here.

Reply:
This method check the `token` parameter, not the `this.CurrentToken`, so the previous switch section covers struct/class/interface.

Reply:
> This method check the token parameter, not the this.CurrentToken - the previous switch section covers struct/class/interface.

Why should this function ever return true if the token is not part of a partial type declaration?

Reply:
This code path is reached for any member of any kind so we are not sure that it's going to be a type or field, etc. We use the next token to disambiguous `partial` being a modifier (and not an identifier), but we don't know if it's a type at that point.